### ISSUE TITLE
Beetle Adventure Racing glitches

### ISSUE THREAD

**Issue by Tasosgemah** [loganmc10/GLupeN64#45]
-Some menu transition screens are missing
-Mount Mayhem stage only show a white fog and nothing else.

**Comment by loganmc10** [comment]
Can you give me some more details? What system are you running this on?

Are you able to provide a save state where the white fog happens? Or a screenshot of thr glitch in the menu?

**Comment by Tasosgemah** [comment]
Here's the Menu screen:

https://s9.postimg.io/6qfyqm5pb/Untitled.jpg

Normally it's not a blank background but there's stuff happening. Other transitions screens are correct.

You don't need a save for Mount Mayhem, just start a new race and select the second level, it's available from the start.

My system is a i5 4670, GTX960 2GB and Windows 7 64bit.

**Comment by loganmc10** [comment]
So I was able to reproduce both these issues with mupen64plus+GLideN64. Are you able to test that to confirm?

This project is a port of mupen64plus+GLideN64 to RetroArch, so if there are issues with the graphics emulation (that are because of this project), then it would be an issue with GLideN64

**Comment by Tasosgemah** [comment]
I tested the game on PJ64 + GlideN64 and it looks pretty much perfect. It was supposed to look perfect anyway since it was one of the focus games for the plugin. 

Maybe it's a problem with the particular revision of GlideN64 used for this core version? Maybe it's a Mupen64 problem? I need to test it with Mupen64+ to be sure, i'll do that later on and report again.

**Comment by fzurita** [comment]
I just tried in mepen64plus-ae and the problem is not there. Maybe it broke in the last few commits? I'm a few commits behind.

**Comment by loganmc10** [comment]
@fzurita I actually just tried it on GLupeN64 for Android and it works there, maybe it only works in GLES mode? That would be strange. Do you have any way to test it with full OpenGL?

**Comment by fzurita** [comment]
No issues on Full OpenGL mode either.

**Comment by loganmc10** [comment]
This really seems like it's got to be an issue with GLideN64. On Fedora, I can get mupen64plus and GLupeN64 plus to work in GLES2 mode, but neither work in OpenGL mode.

**Comment by fzurita** [comment]
Are you sure you are compiling with the right flags for full OpenGL mode?

**Comment by loganmc10** [comment]
On Linux it's pretty straight forward, you just do:
```
cmake -DVEC4_OPT=On -DNOHQ=On -DMUPENPLUSAPI=On ../../src/
```
And you add ```-DGLES2=On``` for GLES2.

I am testing this inside VMware so that's why I'm not 100% sure, I'll test it tonight on a proper machine.

**Comment by loganmc10** [comment]
hm I get this same behaviour with Fedora + Intel Graphics (with mupen64plus+GLideN64)

**Comment by fzurita** [comment]
Interesting, I wonder why the Nvidia Shield TV doesn't have the fame issue when using full OpenGL with mupen64plus-ae.

**Comment by loganmc10** [comment]
It works for me on Windows, but not Fedora 24 (Mesa 12.0.1), it might be some kind of driver thing

**Comment by fzurita** [comment]
Oh, ok. Yeah, it's starting to sound like a driver issue.

**Comment by loganmc10** [comment]
gonetz said it's because that feature needs OpenGL 4.3, which would kind of explain it, I think the Nvidia Shield supports 4.3, but Intel/VMware Mesa only supports 3.3.

However that doesn't really explain why it works with GLES2

**Comment by fzurita** [comment]
GLES 2.0 is based off OpenGL 2.0, so it shouldn't work... Unless there is some #ifdefs in there somewhere.

**Comment by fzurita** [comment]
There is an #ifdef. Search for GL_IMAGE_TEXTURES_SUPPORT.

**Comment by loganmc10** [comment]
Ya but that is not defined for GLES2 or GLES3, but these features work in GLES2 and GLES3, it just seems that they don't work with OpenGL 3.3?

**Comment by orbea** [comment]
This appears to not affect nouveau with mesa-2016.09.04_61e9785 and OpenGL 4.3.

**Comment by Tatsuya79** [comment]
Got this problem too with a similar config as the original poster:
win7 x64 i5-3570K 
Nvidia GTX-770 current stable driver rev

**Comment by Tasosgemah** [comment]
OK so i'm confused.

Is this a problem with this particular version of the plugin?

**Comment by fzurita** [comment]
It seems that a subset of people need there plugin compiled with image texture support disabled.

**Comment by loganmc10** [comment]
I need to do some more testing today but it seems to me that OpenGL 3.3 + GL_IMAGE_TEXTURES_SUPPORT = broken

**Comment by fzurita** [comment]
There should be a way to also configure this option out in addition to the compile time check.

**Comment by loganmc10** [comment]
So yes I tested this and removing GL_IMAGE_TEXTURES_SUPPORT on my machine with OpenGL 3.3 fixes the problem, I've commented/explained this on the upstream issue with GLideN64:
https://github.com/gonetz/GLideN64/issues/1113#issuecomment-246355538

**Comment by loganmc10** [comment]
Although there is still something wrong with GLupeN64 and Windows, it doesn't seem to work there even though OpenGL 4.3 should be supported.

**Comment by loganmc10** [comment]
A fix has been sent to GLideN64:

https://github.com/gonetz/GLideN64/pull/1123

**Comment by Tasosgemah** [comment]
So, forgive me if this is a stupid question, but the fact that the standalone GlideN64 works fine with the game is unrelated?

**Comment by loganmc10** [comment]
GLideN64 has an issue with OpenGL 3.3 and this feature, that is what my change fixed.

Your computer supports OpenGL 4.3, and when you play using mupen64plus+GLideN64 it knows that and everything works good.

RetroArch makes you declare a GL version, so I have it declared at 3.3 since everyone's card supports that (https://github.com/loganmc10/GLupeN64/blob/edf23a091e5f656f2e1039f7d08405469c7fd873/custom/glsm/glsm.c#L2101-L2103)

So basically when you run GLupeN64 it runs at OpenGL 3.3, and therefore this feature is broken without that change I submitted.

**Comment by Tasosgemah** [comment]
Ok now i got it, thanks.

**Comment by loganmc10** [comment]
@Tasosgemah This should be fixed with 1a0ea0176daf1e6a8562345a295e39f859bf24f8, the binary builds have been updated. Can you test it out when you get a chance?

**Comment by Tasosgemah** [comment]
I'd like to, so far i was using the core someone posted on RetroArch forums. Where can i download the latest one? I don't seem to find any links (i'm kinda new here).

**Comment by loganmc10** [comment]
The latest builds can be found here:

http://loganbuildbot.s3-website-us-east-1.amazonaws.com/

**Comment by Tasosgemah** [comment]
Thanks i just tested it.

The menu transitions are good now. Mount Mayhem is playable, however the fog is completely missing now. Normally the fog covers the far parts of the track. Here's a video of how it should look:

https://www.youtube.com/watch?v=RpmJkme8uyo

This is how it looks on standalone GlideN64 too. The reason i picked this game to test was because it was one of the focus games for it. Maybe it indeed needs a newer GL version for these effects?

**Comment by loganmc10** [comment]
Ya it probably needs OpenGL 4.3. I'll talk to the RetroArch people and do a bit of research to see if there is any way to dynamically figure out the user's OpenGL version. Right now RetroArch makes you hardcode it, so it is hard coded to 3.3

**Comment by Tasosgemah** [comment]
Yeah, it seems weird that RetroArch can't just use the best version you have. But as far as i know, this plugin (the standalone GlideN64 that is) was targeted to newer hardware in order to be able to emulate many un-emulated effects that other hle plugins could not. As i tested the standalone with more than 150 games, this is true. So i think using it with only older hardware in mind, it kinda defeats the purpose IMO. 

One last question though, since RetroArch needs a specific GL version to be declared, could it be possible to have two different versions of the core, one for older (3.3) and one for newer (4.3) systems?

**Comment by loganmc10** [comment]
Depth based fog is the only feature that is enabled by allowing OpenGL 4.3 support, so I'd rather have compatibility (it's not just older hardware that uses OpenGL 3.3, anyone with Linux and an Intel Graphics card is on OpenGL 3.3, like me).

I'm also not too keen on supplying 2 different binaries for each platform just for depth based fog support, but I'll look into getting RetroArch to dynamically pick the proper OpenGL version.

In the meantime, if you want to try a version that has OpenGL 4.3 support you can try out this:

[glupen64_libretro.zip](https://github.com/loganmc10/GLupeN64/files/470027/glupen64_libretro.zip)

**Comment by Tasosgemah** [comment]
Works great, thanks. I'll keep that somewhere just for this game.

**Comment by loganmc10** [comment]
So it seems like what RetroArch is doing is actually correct. It is correct to specify an OpenGL version (3.3 in our case), but it **should** give you the highest OpenGL context it can. You'll notice that @orbea didn't have this problem with an Nvidia card on Linux, even though RetroArch requests 3.3, it probably gave a 4.3 context.

For some reason on Windows it doesn't seem to work, it just returns the OpenGL context that was requested, not the highest backwards-compatible context that is available.

Really, this is a problem for RetroArch to figure out, they need to figure out how to properly request the highest OpenGL context possible, with the requested version (3.3) acting as a minimum (ie give an error if your card doesn't support 3.3, because we need at least that, but give me the highest number you possibly can)

**Comment by loganmc10** [comment]
I'll give you a for instance:

On Linux, if I set the OpenGL version to 3.1, it returns an OpenGL context of 3.3 (my max). On Windows, when I request 3.1, it returns 3.1

**Comment by loganmc10** [comment]
@Tasosgemah I've changed the way the OpenGL context is requested and this should be fixed now

http://loganbuildbot.s3-website-us-east-1.amazonaws.com/

**Comment by Tatsuya79** [comment]
Yes, that's working with the buildbot version.
So windows thinks he knows better and we shouldn't tell him anything? :)

**Comment by loganmc10** [comment]
I'm not really sure who to blame, mostly I blame OpenGL, they create ambiguous standards, for instance it says that OpenGL may return:
```
The requested profile of the requested version. Or:
The requested profile of any later version
```

So basically on Linux returns "any later version" and on Windows it returns "the requested version". There are a lot of cases where OpenGL acts one way on Android, and another way on Linux, and another way on Windows. I wish the standards were a little more specific so different implementations acted the same way.

**Comment by Tasosgemah** [comment]
Nice, that was fast :)
